Incorporate `gempyor.distributions` into `gempyor.statistics`

emprzy opened the pull request:
### Describe your changes.

Incorporate the recently added distributions module into `gempyor.statistics`, and add a new `._likelihood()` method to all `Distributions` that calculates the log likelihood of observed ground truth data, given the model data. It is worth noting that integrating this new log likelihood calculation method is not entirely seamless, as the previous implementation in `gempyor.statistics` (`.llik()`) used a `dist_map` to match distributions given in configs to their parameters. My changes leave us in a liminal state between these two, and provide `if`/`else` logic to allow backwards compatibility with `dist_map`, but I think it would be cleaner to move away from `dist_map` entirely. 


### Does this pull request make any user interface changes? If so please describe.

No user interface chagnes


### What does your pull request address? Tag relevant issues.

This pull request addresses GH #583

emprzy committed:
Add `_likelihood` method to `gempyor.distributions`

emprzy committed:
Finish implementation of `DistributionAB._likelihood()`

emprzy committed:
Incorporate `gempyor.distributions` into `gempyor.statistics`

emprzy committed:
Update related tests

emprzy commented:
3 questions going into review:

1. How do we want to approach (if at all) migrating away from `dist_map`?
2. Do we want to try to use NumPy arrays instead of xarrays in `gempyor.statistics`, or would that change be too invasive? I saw there was a comment (presumably left by Joseph) that xarrays are too confusing. It's nice that xarrays have name dimensions, but I agree they are kinda confusing (but maybe I'm just a NumPy girl). 
3. How do we want to go about unit testing `._likelihood()`? I'm not sure of a good way besides individually testing input/expected values for each dist, since all the implementations are so different.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -72,6 +72,24 @@ def _sample_from_generator(
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def _likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Establish distribution-specific likelihood logic."""
+        raise NotImplemented
+
+    def likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """
+        Calculates the log-likelihood of observing data given the model's predictions.
+
+        Args:
+            gt_data: The observed ground truth data.
+            model_data: The data produced by flepiMoP.
+
+        Returns:
+            An array of log-likelihood values.
+        """
+        return self._likelihood(gt_data, model_data)
```
i think this is probably fine as a pattern supporting later flexibility / easier refactor, but: since we don't need a generator or other generic / default setting behavior, why the two defs?

pearsonca replied:
also: naming-wise - this is log likelihood, not likelihood. probably want a different name / offer a log option that defaults to true / etc

emprzy replied:
Very true, not sure why that didn't occur to me.

emprzy replied:
As to the first comment, I'm not really sure. I believe that was the intended implementation @TimothyWillard wanted in the issue (#583 ), but he can either let me know I'm wrong or explain why he wanted the wrapper.

TimothyWillard replied:
Mostly wanted it for consistency with the `_sample_from_generator` abstract method and to leave the door open for us to make changes easily without requiring changes on the implementation side.

pearsonca replied:
k - agree this is generally a good pattern, just checking that we were intentionally pursuing it.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -469,6 +533,14 @@ def _sample_from_generator(
 
+    def _likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Log-likelihood for the beta distribution."""
+        return scipy.stats.beta.logpdf(
+            x=gt_data,
+            a=self.alpha,
+            b=(self.alpha * (1 / np.clip(model_data, 1e-9, 1 - 1e-9) - 1)),
+        )
```
need to double check this, but I think this is likely not the right approach. we should be maintaining a standard deviation while shifting mean - that means moving both a and b.

emprzy replied:
It's certainly possible that I might have flubbed/mangled this during my research (as I certainly did not know how to do it off the top of my head). Perhaps we could introduce `ϕ`, the precision parameter? It would relate alpha and beta in this way:
alpha=mu⋅ϕ
beta=(1−mu)⋅ϕ
But then we'd have to introduce ϕ as a parameter of `BetaDistribution`, which is probably not what we are looking for. Let me know what you learn @pearsonca

pearsonca replied:
I think we do want the precision parameter, at least for beta likelihoods. which means supporting multiple parametrizations of at least the beta distribution

that said - we don't currently support beta likelihoods w/ statistic, so could just put a raise notyetimplemented (or whatever the python TODO error is) here imo.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/statistics.py:
I think the better solution here is to implement the expanded set of likelihood functions as distributions.

emprzy replied:
So just add these to `Distributions.py`? I agree that's a good solution but I just wanted confirmation.

pearsonca reviewed (commented):
This is good start, but there are a few items to tack on.

TimothyWillard commented:
> 1. How do we want to approach (if at all) migrating away from dist_map?

Some of these can be easily replaced by distributions already (like poisson & `norm`/`norm_homoskedastic`), others can be replaced by a new distribution (like `norm_heteroskedastic` similar to normal but the variance scales with the mean). However I do not know for sure what to do about `rmse` or `absolute_error`. This is a bit out of my expertise (@pearsonca or @MacdonaldJoshuaCaleb should weigh in here), but aren't `rmse` & `absolute_error` proportional to normal & laplace distribution log likelihoods, respectively? I also do not think those last two are used in practice.

> 2. Do we want to try to use NumPy arrays instead of xarrays in gempyor.statistics, or would that change be too invasive? I saw there was a comment (presumably left by Joseph) that xarrays are too confusing. It's nice that xarrays have name dimensions, but I agree they are kinda confusing (but maybe I'm just a NumPy girl).

I would say keep them as xarrays for now... Under the hood xarray uses numpy for its array representation so if you design these functions to work on numpy arrays xarray is smart enough to properly apply that to slices of a dataset/dataarray. I think changing this would be better done when we change the output representation from the simulator which starts with the return of `gempyor.seir.steps_SEIR` and then results in changes throughout `gempyor.outcomes` and finally here.

> 3. How do we want to go about unit testing ._likelihood()? I'm not sure of a good way besides individually testing input/expected values for each dist, since all the implementations are so different.

Yeah, that's a fair point and as we've discussed before we do not want to be in the business of testing 3rd party packages. I would say just limit unit testing to where we implement some custom behavior.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -97,6 +115,10 @@ def _sample_from_generator(
 
+    def _likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Log-likelihood for the fixed distribution."""
+        return np.where(gt_data == self.value, 0.0, -np.inf)
```
```suggestion
        return np.where(np.isclose(gt_data, self.value), 0.0, -np.inf)
```

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -127,6 +149,10 @@ def _sample_from_generator(
         """Sampling logic for normal distributions."""
         return rng.normal(loc=self.mu, scale=self.sigma, size=size)
 
+    def _likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Log-likelihood for the normal distribution."""
+        return scipy.stats.norm.logpdf(x=gt_data, loc=model_data, scale=self.sigma)
+
```
It feels odd that you have to declare a mean to this distribution, but that doesn't get used. I'm pointing this out as an example, but I think pretty much all of these likelihoods suffer from this problem. For example, this is a truncated snippet from the `simple_usa_statelevel.yml` config:
```yaml
inference:
  method: emcee
  iterations_per_slot: 100
  do_inference: TRUE
  gt_data_path: model_input/data/generated_gt_data.csv
  statistics:
    incidCase:
      name: incidCase
      sim_var: incidCase
      data_var: incidCase
      resample:
        aggregator: sum
        freq: W-SAT
        skipna: True
      zero_to_one: True
      likelihood: 
        dist: pois
```
this now would become:
```yaml
inference:
  method: emcee
  iterations_per_slot: 100
  do_inference: TRUE
  gt_data_path: model_input/data/generated_gt_data.csv
  statistics:
    incidCase:
      name: incidCase
      sim_var: incidCase
      data_var: incidCase
      resample:
        aggregator: sum
        freq: W-SAT
        skipna: True
      zero_to_one: True
      likelihood: 
        distribution: pois
        lam: 1
```
You now have to add the `lam` and set it to a value so the `pydantic` model validation works, but it's totally unused by the likelihood function. I'm also not sure what the correct fix for this is though... thoughts @emprzy, @pearsonca?

pearsonca replied:
i've been thinking about some adjacent things.

what if we make the parameters optional?

we could write sample to take `**` (at the abstract level, specific args in the implementations), which defaults to None. uses the arguments if available, optional field values if not (and raises if created with no values and asked to sample with no args)

this would work nicely with likelihoods that don't want mean to be specified

emprzy replied:
So we make the parameters part of `**` when you create an instance of a distribution? I like this, but let's take it to the logical extreme; what happens if people have a `Distribution` defined in their (janky) config with no parameters and then the `.sample()` method gets called, an error will come right? Are we ok with that possibility

pearsonca commented:
> Some of these can be easily replaced by distributions already (like poisson & `norm`/`norm_homoskedastic`), others can be replaced by a new distribution (like `norm_heteroskedastic` similar to normal but the variance scales with the mean). However I do not know for sure what to do about `rmse` or `absolute_error`. This is a bit out of my expertise (@pearsonca or @MacdonaldJoshuaCaleb should weigh in here), but aren't `rmse` & `absolute_error` proportional to normal & laplace distribution log likelihoods, respectively? I also do not think those last two are used in practice.

I think RMSE would be ...like a multinormal likelihood? Anywho, I think these are more cases of alternative norms / distance functions, rather than something to try to capture with a likelihood based approach. We want to support other kinds of norms - I think the better answer here is to separate when we're doing a likelihood based norm vs something else, rather than trying to shoehorn these particular ones into a distribution framing.
 
> > 3. How do we want to go about unit testing ._likelihood()? I'm not sure of a good way besides individually testing input/expected values for each dist, since all the implementations are so different.
> 
> Yeah, that's a fair point and as we've discussed before we do not want to be in the business of testing 3rd party packages. I would say just limit unit testing to where we implement some custom behavior.

agree

emprzy committed:
Update flepimop/gempyor_pkg/src/gempyor/distributions.py

Co-authored-by: Timothy Willard <9395586+TimothyWillard@users.noreply.github.com>

emprzy committed:
Change `likelihood` to `loglikelihood`

emprzy committed:
Merge branch 'incorporate-distributions-into-statistics' of https://github.com/HopkinsIDD/flepiMoP into incorporate-distributions-into-statistics

emprzy commented:
@pearsonca @TimothyWillard @MacdonaldJoshuaCaleb See [here](https://gist.github.com/emprzy/9aea4c57aa7c362da24b58f4f68e06c7) an example implementation of the approach we discussed in slack, where there are two different BaseModels for the distributions – one for Sampling and the other for calculating Loglikelihood. Let me know what you think re: our two options right now:

1. Two distinct classes for the two distinct functionalities
2. Using one `DistributionsABC` for both functionalities, making parameter input optional

TimothyWillard commented:
> @pearsonca @TimothyWillard @MacdonaldJoshuaCaleb See [here](https://gist.github.com/emprzy/9aea4c57aa7c362da24b58f4f68e06c7) an example implementation of the approach we discussed in slack, where there are two different BaseModels for the distributions – one for Sampling and the other for calculating Loglikelihood. Let me know what you think re: our two options right now:
> 
> 1. Two distinct classes for the two distinct functionalities
> 2. Using one `DistributionsABC` for both functionalities, making parameter input optional

Now seeing approach (1) implemented I like this a bit more. I think it also gives us more flexibility to tweak and refine likelihoods independent of the rest of the configuration (thinking of #86 for example). Some minor comments before implementation & review in this PR:

1. We don't have to keep the distribution terminology for this, we could call them likelihood terms, and also keep RMSE & absolute error, probably split them out into a different `gempyor.likelihood` module for maintainability.
2. I think the "_sample" and "_loglike" suffixes aren't needed since this are distinct concepts and users will know which one is being used depending on the configuration section being worked in.
3. I like the template method pattern that we're rolling with (`log_likelihood` being provided by `LogLikelihoodTermABC` calling `_log_likelihood` provided by implementations). I think this gives us a lot of flexibility to handle new features in the future.

emprzy committed:
Add a new `gempyor.likelihoods` module

Split distributions used for sampling and llik calculation into distinct modules.

emprzy committed:
Refactor tests for `gempyor.statistics`

emprzy committed:
Build testing suite for new `gempyor.likelihoods` module

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+        raise NotImplemented
+
+    def loglikelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """
+        Calculates the log-likelihood of observing data given the model's predictions.
+
+        Args:
+            gt_data: The observed ground truth data.
+            model_data: The data produced by flepiMoP.
+
+        Returns:
+            An array of log-likelihood values.
+        """
```
whatever the proper docstring syntax here, let's have an details noting that:
 - `gt_data`, `model_data` must be the same shape (leave aside for the moment any actual enforcement of this, however)
 - the caller is responsible for ensuring that gt_data and model_data align (i.e. this does nothing to deal with observation times versus model time)

emprzy replied:
This is a good point, though do we think there would be failures earlier than this code if `gt_data` and `model_data` do not have matching dimensions? Users aren't going to interact with the `._likelihood()` method directly .

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+    def _loglikelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Log-likelihood calculations for uniform distributions."""
+        loc = model_data - ((self.high - self.low) / 2.0)
+        scale = self.high - self.low
```
```suggestion
        loc = model_data - (scale / 2.0)
```

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+    """
+
+    distribution: Literal["uniform"] = "uniform"
+    low: EvaledFloat
+    high: EvaledFloat
```
```suggestion
    scale: EvaledFloat = Float(..., gt = 0)
```

pearsonca replied:
It's a little weird to think about a uniform likelihood, but when we do, we're actually talking about the *width* of the interval, not the high and low.

Mathematical aside: part of the reason we know this is one parameter is because we never use anything except high - low.

emprzy replied:
We can't remove `scale` from within the `._loglikelihood()` method because the `scipy.stats.uniform.logpdf()` call uses it. But also my opinion is that we should get rid of uniform as a distribution available for llik calculation.

emprzy replied:
Note for posterity: removed uniform as a distribution available for likelihood

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+
+class TruncatedNormalLoglikelihood(LoglikelihoodABC):
+    """
+    Represents a truncated normal distribution for calculating log-likelihood.
```
need to remind people that the interpretation of a, b here is different from what's going on w/ trunc normal *sampling*.

In sampling, a and b are the absolute sample limits. Here they should be the relative-offsets-to-mu sample limits.

pearsonca replied:
hmm, actually I think there's a lot more going on here with how we want to allow a truncated normal likelihood that's going to take some mathematical care.

for *now*, we shouldn't support this. we can revisit it *if* people request it in the future.

emprzy replied:
i agree with this take, and feel the same way about uniform for llik. let me know if you disagree @TimothyWillard @pearsonca

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+
+    def _loglikelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Log-likelihood calculations for binomial distributions."""
+        return scipy.stats.binom.logpmf(k=gt_data, n=self.n, p=np.clip(model_data, 0, 1))
```
Unclear we want to clip here - if the observation model wants to clip, that seems fine, but the likelihood calculation should error if offered p > 1.

emprzy replied:
Ooo yeah I agree about this. I'll change this to have an error point instead of the clip

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/likelihoods.py:
```diff
@@ -0,0 +1,294 @@
+        )
+
+
+class AbsoluteErrorLoglikelihood(LoglikelihoodABC):
```
Abs Error isn't a likelihood - we need an additional layer of distance or norm or target evaluation here - so like TargetABC => LikelihoodABC => [Distributions]ABC. But also TargetABC => AbsoluteError. Ibid for RMSE

emprzy replied:
Definitely hear where you are coming from, though I think a better solution to this dilemma would not to create more parent ABCs, but to create a better name for the `LogLikelihoodABC` class that is more inclusive of these two "flavors" of error calculation. Like, just renaming `LogLikelihoodABC` to `ErrorMetricABC`,  `MetricABC`, `ObjectiveABC`, or something like that? Keeps things simple and clean. @pearsonca

emprzy replied:
`ObjectiveFunctionABC` ...? Any takers...?

pearsonca commented on flepimop/gempyor_pkg/tests/distributions/test_distributions_common.py:
```diff
@@ -18,6 +19,11 @@ def _sample_from_generator(
         """A fake sampling implementation."""
         return rng.random(size)
 
+    def _likelihood(self, gt_data: npt.NDArray, model_data: npt.NDArray) -> npt.NDArray:
+        """Dummy likelihood implementation."""
+        # dummy, predictable output
+        return -((gt_data - model_data) ** 2)
+
```
vestigal?

emprzy replied:
vestigal why?

pearsonca replied:
because we aren't doing likelihoods in distributions anymore.

emprzy replied:
oops, totally, yes.

pearsonca commented on flepimop/gempyor_pkg/tests/likelihoods/test_absolute_error_likelihood_class.py:
Hmm - aren't we repeating the test-them-all error here?

emprzy replied:
this is the same framework we used for testing `gempyor.distributions` (one test file per distribution). or is there something specific about this file that you think is repetitive..?

pearsonca reviewed (changes requested):
Seems like progress, but some lingering bits from switch to split distribution vs likelihood. There are also some issues with the likelihood specifications, but where I think we can put off some implementations until they are actually needed.

emprzy committed:
Implement most recent PR suggestions [skip ci]

Change `gempyor.likelihoods` to be `gempyor.objective_functions`

emprzy commented:
In my most recent commit ([32fc04c](https://github.com/HopkinsIDD/flepiMoP/pull/595/commits/32fc04c536ec48e4b54112dc2e7c54dbaca5ee96)) I implement some of the minute suggestions from @pearsonca 's last review, and I also establish a new naming convention for what was previously called `gempyor.likelihoods` to hopefully be more inclusive of RMSE and AbsErr without having to create more parent ABCs. Let me know what y'all think of that implementation. 
@pearsonca @TimothyWillard 

I'll update things like tests and docstring documentation once I get the green light on this!

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/objective_functions.py:
```diff
@@ -0,0 +1,279 @@
+
+    @abstractmethod
+    def _error_metric_calculation(
+        self, gt_data: npt.NDArray, model_data: npt.NDArray
```
note for later: i think `gt_data` and `model_data`, as `npt.NDArray` is fine for now, but we'll want to revisit this for more generic meaning of objective functions. for example, thinking about optimal control approaches: there wouldn't be a ground truth data - more like a cost per event data, and model data would be time series of not just health states.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/objective_functions.py:
+1 re objective functions - great choice.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/objective_functions.py:
```diff
@@ -0,0 +1,279 @@
+        self, gt_data: npt.NDArray, model_data: npt.NDArray
+    ) -> npt.NDArray:
+        """Log-likelihood calculations for lognormal distributions."""
+        return scipy.stats.lognorm.logpdf(x=gt_data, s=self.sdlog, scale=model_data)
```
have to double check this - pretty sure its not right, but i'm not sure what the correct version is.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/objective_functions.py:
```diff
@@ -0,0 +1,279 @@
+    """
+    Calculates an error metric using the sum of absolute errors..
+
+    The final score is calculated as -log(sum_of_absolute_errors).
```
was the previous version also log scale? this is a bit weird, imo

emprzy replied:
note for posterity: yes the previous version was a log scale, leaving it as such for now. can be changed in the future

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/statistics.py:
hmmmm, what are we doing with the zero-to-one shifts?

pearsonca reviewed (commented):
I need to do some closer look at some of the likelihood calculations, but perhaps something to work with ahead of that.

emprzy committed:
Final updates

Finish testing suite, update documentation,

emprzy committed:
Fixing expected float values in docstring examples of `objective_functions.py`

emprzy commented:
Only thing left here is for @pearsonca to review whether or not scipy is doing what we want!

pearsonca committed:
tweaks to distributions

pearsonca committed:
review all objective functions

emprzy committed:
Update tests surrounding `distributions.py`

emprzy committed:
Update tests surrounding `objective_functions.py`

Also, disable RMSE and Abs.Error as objective functions for now

emprzy commented:
Note that we have had to disable `RMSE` and `AbsoluteError` as objective functions due to the fact that `gempyor.statistics::llik()` is not presently able to handle scalar returns from an error metric calculation. Created #612 to address

emprzy committed:
Merge branch 'dev' into incorporate-distributions-into-statistics
